aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -1221,44 +1330,35 @@ Tensor sparse_compressed_to_flipped(
 }
 
 Tensor sparse_compressed_to_sparse_csr(const Tensor& self, c10::optional<int64_t> dense_dim_opt) {
-  if (dense_dim_opt.has_value()) {
-    AT_ERROR("sparse_compressed_to_sparse_csr conversion does not support specifying number of dense dimensions");
-  }
+  auto layout_to = kSparseCsr;
+  TORCH_CHECK(self.layout() != layout_to, "sparse_compressed_to_sparse_csr: same input and output layouts are not supported");
```

Comment:
Reading further, I see it will work because the same layout case will be handled by `torch.to_sparse` implementation one step above.

Perhaps this should be `TORCH_INTERNAL_ASSERT` as a user should not be able to trigger this exception. Is this correct?

Reply:
True, the whole point of this PR is to have a method like `to_sparse` that will handle the case when input and output layouts the same, and then another method `_to_sparse` that would handle the rest of the cases, and have an expliciit backward function.

Still, methods like `_to_sparse` are visible to user, for example following works:
```python
x = torch.sparse_coo_tensor(indices=[[0, 1], [0, 1]], values=[1, 1])
print(x._to_sparse())
```
and will report a runtime error due to a check like the one above.  So, I'm not sure shall this check be changed to an assert?

Reply:
Although `_to_sparse` (among many other private methods) is visible to users, it is a private method that users are expected not to use. So, the given example is not valid for designing this interface, imho.

If the assert is triggered by a user via some public pytroch function/method, then it is likely a pytorch bug and ought to be fixed.

Reply:
Ok, changed alike checks to `TORCH_INTERNAL_ASSERT`.